### PR TITLE
fix: github issue #4439: case insensitive compare of file paths.

### LINKED ISSUE #4439: Rollup build fails on windows: SyntaxError: Cannot use import statement outside a module
### Rollup Version

2.70.0

### Operating System (or Browser)

Windows 11

### Node Version (if applicable)

v16.14.0

### Link To Reproduction

https://github.com/Vtec234/vscode-lean4

### Expected Behaviour

The following steps should build without error - it builds without error on Linux, but not on Windows:

1. git clone https://github.com/Vtec234/vscode-lean4
2. git checkout infoview-hooks
3. npm install
4. npx lerna bootstrap
5. npm run build

This even passes in an MSYS2 environment on the same windows machine using the exact same npm and nodejs binaries.  

### Actual Behaviour

On Windows I get this error:
```
[!] SyntaxError: Cannot use import statement outside a module
d:\Temp\vtec234\vscode-lean4\lean4-infoview\rollup.config.js:1
import commonjs from '@rollup/plugin-commonjs';
^^^^^^
```

On debugging the error is originating here:

```
SyntaxError: Cannot use import statement outside a module
  at Object.compileFunction (node:vm:352:18)
  at wrapSafe (node:internal/modules/cjs/loader:1032:15)
  at Module._compile (node:internal/modules/cjs/loader:1067:27)
  at Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
  at Object.require.extensions.<computed> [as .js] (c:\\Users\\clovett\\AppData\\Roaming\\npm\\node_modules\\rollup\\dist\\shared\\loadConfigFile.js:617:13)
  at Module.load (node:internal/modules/cjs/loader:981:32)
  at Function.Module._load (node:internal/modules/cjs/loader:822:12)
  at Module.require (node:internal/modules/cjs/loader:1005:19)
  at require (node:internal/modules/cjs/helpers:102:18)
  at loadConfigFromBundledFile (c:\\Users\\clovett\\AppData\\Roaming\\npm\\node_modules\\rollup\\dist\\shared\\loadConfigFile.js:622:42)
```

### PR DESCRIPTION
**Pull request by lovettchris**
This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

resolves #4439

### Description

On windows the string returned from `await fs.realpath(fileName)` might contain lower case drive letter while the string returned from `path.resolve(__dirname, ...)` can return a different case, which causes direct string comparisons using `===` to fail.  This fix adds a win32 specific implementation of a helper function `isSamePath` that ensures the comparison is case insensitive on a win32 platform.

### CODE REVIEW

**Review by lukastaegert — COMMENTED** [review]
Do you think we can add a test for this, after all, tests are also running on Windows. There is a special test file just for `loadConfigFile`, or maybe a test case in the tests/cli folder?

File: scripts/load-perf-config.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ exports.loadPerfConfig = async () => {
 function loadConfigFromCode(code) {
 	const defaultLoader = require.extensions['.js'];
 	require.extensions['.js'] = (module, filename) => {
-		if (filename === configFile) {
+		if (rollup.isSamePath(filename, configFile)) {
```
[the remark this answers is not in the document]
Comment (lovettchris) [reply scripts/load-perf-config.js:31]: Changed this to use a local copy of the isSamePath function.

File: package-lock.json [hunk 2]
```diff
@@ -1,8 +1,6571 @@
 {
   "name": "rollup",
   "version": "2.70.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
```
[the remark this answers is not in the document]
Comment (lovettchris) [reply package-lock.json:4]: yeah, sorry about that, didn't mean to check this in.

[the remark this answers is not in the document]
Comment (lovettchris) [reply package-lock.json]: Reverted

**Comment by lovettchris** [comment]
> Do you think we can add a test for this, after all, tests are also running on Windows. There is a special test file just for `loadConfigFile`, or maybe a test case in the tests/cli folder?

yeah, I've been having a similar conversation over here: https://github.com/rollup/plugins/pull/1134.  I can certainly use that trick to make all the rollup tests fails in the same way.  Just not sure how to do that in the CI VM...

**Comment by lovettchris** [comment]
Note there may be one more error lerking in the code because when I force the default drive letter to be wrong (using the tricks I mentioned in https://github.com/rollup/plugins/pull/1134, then `npm run build` also fails as follows:

image

PS: and this is after I patched the older version 2.67.1 of `node_modules\rollup`.  Interesting that rollup is used to build rollup :-)  Must have been an interesting bootstrapping process...

**Comment by dnalborczyk** [comment]
> On windows the string returned from `await fs.realpath(fileName)` might contain lower case drive letter while the string returned from `path.resolve(__dirname, ...)` can return a different case, which causes direct string comparisons using `===` to fail.

I wonder if this is a node.js bug?

**Comment by dnalborczyk** [comment]
btw @lukastaegert 

this logic: https://github.com/rollup/rollup/pull/4440/files#diff-167a57161916452cc80dedbc71446dc3d1fdd43e8db6304525f7990b4759f57dR118 

`requiredFileName === resolvedFileName`

is also responsible for this issue: https://github.com/rollup/rollup/issues/4465

but in this case they are both different paths, one is the 'real path', and the other the mounted path (subst). this PR will not address this. I wonder if there's a better solution to fix both issues?

I also tried to figure out why this comparison is in place, but didn't get anywhere. the entire logic of reading just a config file seems to be overly complicated and convoluted on first sight. (not judging). I imagine it grew organically without much refactoring and one of the obstacles is also trying to read many formats js (as cjs and esm), cjs, mjs (and ts?) configs. but it still feels like the entire logic needs a bit of an overhaul.

it also appears that this is also a good candidate for a v3 refactoring (regarding proper cjs and esm parity with node.js, which it looks like rollup is doing it's own thing so to speak).

**Comment by lukastaegert** [comment]
Closing this in favor of #4501, which should solve the same issue but I think is a more general solution.

**Pull request closed without merging**